src/pulse-tracing-qvm.lisp

```diff
@@ -0,0 +1,198 @@
+(in-package :qvm)
+
+;;; TODO is there a better name?
+;;; the event can either be tx or rx
+(defstruct pulse-event
+  instruction
+  start-time
+  end-time
+  frame-state)
+
+;;; TODO move this to cl-quil ast code?
+(defun pulse-op-frame (instr)
+  (etypecase instr
+    (quil:pulse (quil:pulse-frame instr))
+    (quil:capture (quil:capture-frame instr))
+    (quil:raw-capture (quil:raw-capture-frame instr))))
+
+(defun pulse-event-frame (event)
+  "The frame associated with a pulse event."
+  (let ((instr (pulse-event-instruction event)))
+    (pulse-op-frame instr)))
+
+(defun pulse-event-duration (pulse-event)
+  "The duration of a pulse event."
+  (- (pulse-event-end-time pulse-event)
+     (pulse-event-start-time pulse-event)))
+
+;;; A structure tracking the state of a specific frame.
+(defstruct frame-state
+  (phase 0.0d0
+   :type real)
+  (scale 1.0d0
+   :type real)
+  (frequency (error "Frequency must be defined.")
+   :type real))
+
+(defparameter *default-frame-frequency* 1.0 ; TODO get a better default
+  "The default frame frequency.")
+
+(defparameter *initial-pulse-event-log-length* 100
+  "The initial length of the pulse tracing QVM's event log.")
+
+;;; This is pretty barebones, but does make some claims about what is important
+;;; to track. Namely, qubit local time and frame states. We also update a log,
+;;; although TODO in principle this could be more generic (e.g. provide some
+;;; sort of "event consumer" callbacks).
+(defclass pulse-tracing-qvm (classical-memory-mixin)
+  ((sample-rate :initarg :sample-rate
+                :accessor sample-rate
+                :initform (error "A sample rate must be specified.")
+                :documentation "The sample rate in Hz for waveform generation.")
+   (local-clocks :initarg :local-clocks
+                 :accessor local-clocks
+                 :initform (make-hash-table)
+                 :documentation "A table mapping qubit indices to the time of their last activity.")
+   (frame-states :initarg :frame-states
+                 :accessor frame-states
+                 :initform (make-hash-table :test 'equalp) ; TODO move to alist so we can use a custom equality check?
+                 :documentation "A table mapping frames to their active states.")
+   (pulse-event-log :initarg :log
+                    :accessor pulse-event-log
+                    :initform  (make-array *initial-pulse-event-log-length*
+                                           :fill-pointer 0
+                                           :adjustable t)
+                    :documentation "A log, in reverse chronological order, of observed pulse events."))
```

Comment:
Is it reverse chronological order if you're using `vector-push-extend` to push onto the end of the vector?

Reply:
Nice to know that somebody around here is reading the docs 😛

Reply:
Only way I can keep up...